### PR TITLE
Explain default trait object liftime bounds.

### LINKED ISSUE #31137: Document "Default object lifetime bounds"
"Default object lifetime bounds" rust-lang/rfcs/pull/1156

This should be mentioned in the language reference.

### PR DESCRIPTION
**Pull request by steveklabnik**
Fixes #31137 

I feel a little weird about how short this is. It's _really_ hard to not just say the exact rule, about `T: 'a`. It feels so arbitrary here, though.

r? @aturon @nikomatsakis

### CODE REVIEW

**Comment by nikomatsakis** [comment]
I'm not sure if this gets at the essence of the rule. For example, it doesn't example what happens with `Box<Trait>`. I think the rule is kind of aptly summarized as consulting, in order:

1. Lifetime bounds on the trait, if any. For example, `trait Foo: 'static` or `trait Bar<'a>: 'a`.
2. Lifetime of enclosing reference, if any. So, `&Trait` => `&'a (Trait+'a)`
3. Else, `'static`.

**Comment by nikomatsakis** [comment]
Though I think there may be some subtleties about `Ref<'a, Trait>` -- I can't remember where we ended up on that (did we modify that case when we tweaked the default object lifetime rule?)

**Comment by steveklabnik** [comment]
@nikomatsakis updated, what do you think of this attempt?

**Comment by nikomatsakis** [comment]
@steveklabnik seems better. Something still gives me pause. I feel like the way the importance of the cases is the reverse of they are listed. That is, another way to think of it is "trait objects cannot contain borrowed data unless one of these cases is met":

- the trait declares a bound, e.g. `trait Foo<'a>: 'a`
- the object type appears underneath a reference, e.g., `&Foo`
- the object type has an explicit bound `Box<Foo+'a>`

It's just a matter of emphasis though. Thoughts?

**Comment by steveklabnik** [comment]
Ah, I like that.

**Comment by steveklabnik** [comment]
@nikomatsakis r?

File: src/doc/reference.md [hunk 1]
```diff
@@ -3603,6 +3603,41 @@ fn main() {
 In this example, the trait `Printable` occurs as a trait object in both the
 type signature of `print`, and the cast expression in `main`.
 
+Trait objects may contain references, and so those references will need a
+lifetime. If the trait contains some sort of lifetime bound, like this:
```
Comment (nikomatsakis) [inline src/doc/reference.md]: feels weird to me that this sentence just... ends (and doesn't resume after the code snippet)

File: src/doc/reference.md [hunk 2]
```diff
@@ -3603,6 +3603,41 @@ fn main() {
...
+
+If there’s no bound, then the default lifetime is the same as the pointer that
+the trait object is behind:
+
+```rust,ignore
+// Some trait like this...
+trait Bar { }
+
+// ... means these are the same:
+&Bar
```
Comment (nikomatsakis) [inline src/doc/reference.md]: maybe write &'a Bar here?

File: src/doc/reference.md [hunk 3]
```diff
@@ -3603,6 +3603,41 @@ fn main() {
...
+
+In any other case, the default is `'static`:
+
+```rust,ignore
+// Some trait like this...
+trait Bar { }
+
+// ... means these are the same:
+Box<Bar>
+Box<Bar + 'static>
```
Comment (nikomatsakis) [inline src/doc/reference.md]: one case you may want to clarify is that `&Box<Bar>` is same as `&Box<Bar+'static>` -- i.e., it's the *innermost* case that matters

Comment (nikomatsakis) [inline src/doc/reference.md]: there is technically I think one other rule, not sure how much detail we want to go into though. I think if you have `struct Ref<'a, T: ?Sized>` and you do `Ref<'a, Trait>` you get `Ref<'a, Trait+'a>` by default. Unless we changed that in the follow-up RFC. :P I guess I should go double check.

**Comment by steveklabnik** [comment]
@nikomatsakis after a very long time, nits addressed, i think.

**Comment by solson** [comment]
Why was this closed without merging? https://github.com/rust-lang/rust/issues/31137 is still open as well.

**Comment by steveklabnik** [comment]
I believe I was cleaning up my branches and may have accidentally deleted it. It was also sitting in the queue for a very, very long time...

**Comment by solson** [comment]
@steveklabnik Would it be okay if I took your changes from this PR and submitted them as a new PR?

**Pull request closed without merging**